src/shims/unix/fs.rs

```diff
@@ -1253,15 +1253,17 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                     }
                     _ => unreachable!(),
                 }
-
-                let result_place = this.deref_pointer(result_op)?;
+                let result_place =
+                    this.deref_pointer_as(result_op, this.machine.layouts.mut_raw_ptr)?;
                 this.write_scalar(this.read_scalar(entry_op)?, &result_place)?;
 
                 Scalar::from_i32(0)
             }
             None => {
                 // end of stream: return 0, assign *result=NULL
-                this.write_null(&this.deref_pointer(result_op)?)?;
+                let result_place =
+                    this.deref_pointer_as(result_op, this.machine.layouts.mut_raw_ptr)?;
```

Comment:
Please move this to the top of the function (just after `let dirp`) to avoid duplicating it.